File: csharp/matching-brackets/MatchingBrackets.cs
public static class MatchingBrackets
{
    public static bool IsPaired(string input)
    {
        Stack<char> stack = new();
        foreach (char c in input)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                stack.Push(c);
            }
            else if (c == ')' || c == ']' || c == '}')
            {
                if (stack.Count == 0) return false;
                char top = stack.Peek();
                if ((c == ')' && top != '(') ||
                    (c == ']' && top != '[') ||
                    (c == '}' && top != '{'))
                {
                    return false;
                }
                stack.Pop();
            }
        }
        return stack.Count == 0;
    }
}

File: typescript/matching-brackets/matching-brackets.ts
type ClosingBracket = '}' | ']' | ')';
type OpeningBracket = '{' | '[' | '(';

export function isPaired(input: string): boolean {
  const stack: OpeningBracket[] = [];
  const pairs: Record<ClosingBracket, OpeningBracket> = {
    '}': '{',
    ']': '[',
    ')': '(',
  };
  const openingBrackets = new Set<string>(['{', '[', '(']);
  for (const char of input) {
    if (openingBrackets.has(char)){
      stack.push(char as OpeningBracket);
    } else if (char in pairs){
      const closing = char as ClosingBracket
      if (stack.pop() !== pairs[closing]){
        return false;
        }
    }
  }
  return stack.length === 0;
}
